File: ui/src/server/rate-limit.ts
// Rate limiting — Redis-backed, so it survives a restart and holds across
// instances. A fixed window (INCR + EXPIRE on first hit) rather than a sliding
// one: it can let through up to 2x the limit across a window boundary, which is
// the right trade for a brute-force brake where the goal is "thousands of
// guesses become dozens", not exact accounting.
import { getRedis } from './db/redis'

export interface RateLimitResult {
  /** False when the caller should be refused. */
  ok: boolean
  /** Seconds until the window resets — send as Retry-After. */
  retryAfter: number
}

/** Count one hit against `key`. Fails OPEN if Redis is unreachable: a limiter
 *  outage must not lock everyone out of their own deployment. */
export async function rateLimit(key: string, limit: number, windowSeconds: number): Promise<RateLimitResult> {
  try {
    const redis = getRedis()
    const k = `talaria:rl:${key}`
    const count = await redis.incr(k)
    if (count === 1) await redis.expire(k, windowSeconds)
    if (count <= limit) return { ok: true, retryAfter: 0 }
    const ttl = await redis.ttl(k)
    return { ok: false, retryAfter: ttl > 0 ? ttl : windowSeconds }
  } catch {
    return { ok: true, retryAfter: 0 }
  }
}

/** Drop the counter for `key` (a successful login shouldn't leave the failure
 *  budget spent). */
export async function rateLimitReset(key: string): Promise<void> {
  await getRedis()
    .del(`talaria:rl:${key}`)
    .catch(() => {})
}

/**
 * The client's address, for limiting purposes only.
 *
 * X-Forwarded-For is caller-supplied and trivially rotated, so trusting it
 * blindly turns a per-IP limit into no limit at all — which is exactly what the
 * previous implementation did. It is only honored when the operator states that
 * a proxy really is in front (TALARIA_TRUST_PROXY), because only then is the
 * left-most entry something the client can't freely invent.
 *
 * With no trusted proxy this returns a constant, and callers should lean on the
 * per-identity limit instead. That is deliberate: a wrong IP is worse than no
 * IP, because it silently partitions the counter per attacker-chosen value.
 */
export function clientIp(request: Request): string {
  const trust = (process.env.TALARIA_TRUST_PROXY ?? '').trim().toLowerCase()
  if (trust === '' || trust === '0' || trust === 'false' || trust === 'no') return 'direct'
  const xff = request.headers.get('x-forwarded-for')
  const first = xff?.split(',')[0]?.trim()
  return first || request.headers.get('x-real-ip')?.trim() || 'direct'
}

File: api/src/ratelimit.rs
// Rate limiting — port of ui/src/server/rate-limit.ts. Redis-backed so a
// counter survives a restart and holds across BOTH runtimes during the
// migration (same key space, same window arithmetic). A fixed window (INCR +
// EXPIRE on first hit), not a sliding one: it can let through up to 2x the
// limit across a window boundary, which is the right trade for a brake whose
// goal is "thousands become dozens", not exact accounting.
//
// Unit-tested now; wired into the chat-completions route's rpm brake.

use redis::AsyncCommands;
use redis::aio::ConnectionManager;

pub struct RateLimitResult {
    /// False when the caller should be refused.
    pub ok: bool,
    /// Seconds until the window resets — send as Retry-After.
    pub retry_after: i64,
}

/// The Redis key for a limiter name. Byte-identical to the TS spelling so the
/// two runtimes share one counter, whatever is serving the route today.
pub fn redis_key(key: &str) -> String {
    format!("talaria:rl:{key}")
}

/// Count one hit against `key`. Fails OPEN if Redis is unreachable: a limiter
/// outage must not lock everyone out of their own deployment.
pub async fn rate_limit(
    redis: &mut ConnectionManager,
    key: &str,
    limit: i64,
    window_seconds: i64,
) -> RateLimitResult {
    let k = redis_key(key);
    let counted = async {
        let count: i64 = redis.incr(&k, 1).await?;
        if count == 1 {
            let _: bool = redis.expire(&k, window_seconds).await?;
        }
        if count <= limit {
            return Ok::<_, redis::RedisError>(RateLimitResult {
                ok: true,
                retry_after: 0,
            });
        }
        let ttl: i64 = redis.ttl(&k).await?;
        Ok(RateLimitResult {
            ok: false,
            retry_after: if ttl > 0 { ttl } else { window_seconds },
        })
    };
    match counted.await {
        Ok(r) => r,
        Err(_) => RateLimitResult {
            ok: true,
            retry_after: 0,
        },
    }
}

/// Drop the counter for `key` (a success shouldn't leave the budget spent).
/// No Rust caller yet — the TS routes that use it are a later batch.
#[allow(dead_code)]
pub async fn rate_limit_reset(redis: &mut ConnectionManager, key: &str) {
    let _: Result<i64, _> = redis.del(redis_key(key)).await;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn key_shape_matches_the_ts_limiter() {
        // The gateway chat route's key is `gw:key:{keyId}` in TS
        // (llm.v1.chat.completions.ts); the shared counter depends on this
        // exact string.
        assert_eq!(redis_key("gw:key:abc"), "talaria:rl:gw:key:abc");
    }
}
